etna/feedback/views.py

```diff
@@ -0,0 +1,197 @@
+from django.views.generic import FormView, TemplateView
+
+from wagtail.admin.auth import permission_denied
+from wagtail.admin.views.reports import ReportView
+from wagtail.models import Revision
+
+from etna.feedback.forms import FeedbackForm
+from etna.feedback.models import FeedbackPrompt, FeedbackSubmission
+
+logger = logging.getLogger(__name__)
+
+
+class VersionedFeedbackViewMixin:
+    """
+    Mixin for view classes that are initialized from a prompt_id/version
+    combination, which determine options and messaging shown to the user.
+    """
+
+    def setup(
+        self,
+        request: HttpRequest,
+        prompt_id: uuid.UUID,
+        version: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        super().setup(request, **kwargs)
+        self.prompt_id = prompt_id
+        self.version = version
+        self.is_ajax = request.POST.get("is_ajax", "false") == "true"
+        if version:
+            self.prompt_revision = get_object_or_404(
+                Revision,
+                content_type=ContentType.objects.get_for_model(FeedbackPrompt),
+                id=version,
+            )
+            self.prompt = self.prompt_revision.as_object()
+            if self.prompt.public_id != prompt_id:
+                raise Http404("Bad prompt_id / version combination.")
+        else:
+            self.prompt = get_object_or_404(
+                FeedbackPrompt.objects.all().select_related("latest_revision"),
+                public_id=prompt_id,
+            )
+            self.prompt_revision = self.prompt.latest_revision
+
+
+@method_decorator(csrf_exempt, name="dispatch")
+class FeedbackSubmitView(VersionedFeedbackViewMixin, FormView):
+    template_name = "feedback/submit.html"
+    form_class = FeedbackForm
+
+    def get_form_kwargs(self) -> Dict[str, Any]:
+        kwargs = super().get_form_kwargs()
+        request = self.request
+        kwargs.update(
+            response_options=self.prompt.response_options,
+            referer=request.META.get("HTTP_REFERER", ""),
+            url_path_must_match_referer=request.resolver_match.namespace != "feedback",
+        )
+        return kwargs
+
+    def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
+        return super().get_context_data(prompt=self.prompt, **kwargs)
+
+    def form_valid(self, form: Form) -> HttpResponse:
+        # Initialize submission
+        obj = FeedbackSubmission(
+            prompt=self.prompt,
+            prompt_text=self.prompt.text,
+            prompt_revision=self.prompt_revision,
+        )
+
+        # Set field values from form data
+        for name, value in form.cleaned_data.items():
+            setattr(obj, name, value)
+
+        # Update fields from request data
+        if self.request.user.is_authenticated:
+            obj.user = self.request.user
+
+        # Save submission to the DB
+        obj.save()
+
+        if self.is_ajax:
+            return JsonResponse({"id": str(obj.public_id)})
```

Comment:
Question: I'm not familiar with AJAX. Its here in a couple of places. What steps do I take in order to test this?

Reply:
When you have javascript enabled, and 'is_ajax' parameter is added to the `POST` data before posting to this view. 

This view simply looks for that parameter in `setup()` and sets `self.is_ajax` based on the presence.

You can tell this is working, because the JS in the `prompt.html` template is able to decode the JSON and render the success message.